File: src/lib/glbParser.js
/**
 * lib/glbParser.js
 *
 * Parses a single mago3d-tiler .glb point-cloud file and returns
 * an array of geodetic points: Array<{ lon: number, lat: number, h: number }>
 *
 * No Cesium, no React, no browser APIs beyond ArrayBuffer / DataView / TextDecoder.
 * Safe to run in a Web Worker or Node.js.
 *
 * ── mago3d-tiler GLB format (verified against real files) ───────────────────
 *
 *  Node 0  "RootNode"
 *    translation: [tx, ty, tz]   ← ECEF position in glTF Y-up space
 *    children: [1]
 *
 *  Node 1  "PointCloudNode"
 *    matrix: [16 floats, column-major]  ← uniform scale + local offset
 *    mesh: 0
 *
 *  POSITION accessor:
 *    componentType: 5123 (uint16), normalized: true  → raw / 65535 = 0..1
 *    bufferView.byteStride: 8  (3 × uint16 = 6 bytes + 2 bytes padding)
 *
 *  Transform pipeline:
 *    1. lx,ly,lz = uint16 / 65535
 *    2. Apply PointCloudNode matrix M (column-major 4×4):
 *         wx = M[0]*lx + M[4]*ly + M[ 8]*lz + M[12]
 *         wy = M[1]*lx + M[5]*ly + M[ 9]*lz + M[13]
 *         wz = M[2]*lx + M[6]*ly + M[10]*lz + M[14]
 *    3. Add RootNode translation:
 *         gx = wx + rt[0],  gy = wy + rt[1],  gz = wz + rt[2]
 *    4. glTF Y-up → ECEF Z-up:
 *         ecef_x = gx,  ecef_y = -gz,  ecef_z = gy
 *    5. ECEF → (lon°, lat°, h m)  via Bowring iterative method
 *
 *  Verified: lon ≈ 127.006°, lat ≈ 36.908°, h ≈ 23 m  ✓ (Korea site)
 */

// ── WGS-84 constants ───────────────────────────────────────────────────────
const WGS84_A  = 6378137.0
const WGS84_E2 = 0.00669437999014

/**
 * Convert ECEF Cartesian (x, y, z) metres to geodetic (lon°, lat°, h m).
 * Uses Bowring's iterative method (10 iterations — converges to mm accuracy).
 */
function ecefToGeodetic(ex, ey, ez) {
  const p   = Math.sqrt(ex * ex + ey * ey)
  let lat   = Math.atan2(ez, p * (1 - WGS84_E2))
  for (let i = 0; i < 10; i++) {
    const s = Math.sin(lat)
    const N = WGS84_A / Math.sqrt(1 - WGS84_E2 * s * s)
    lat = Math.atan2(ez + WGS84_E2 * N * s, p)
  }
  const s = Math.sin(lat)
  const c = Math.cos(lat)
  const N = WGS84_A / Math.sqrt(1 - WGS84_E2 * s * s)
  return {
    lon: Math.atan2(ey, ex) * (180 / Math.PI),
    lat: lat * (180 / Math.PI),
    h:   Math.abs(c) > 1e-9 ? p / c - N : Math.abs(ez) / s - N * (1 - WGS84_E2),
  }
}

/**
 * Parse one mago3d-tiler .glb ArrayBuffer.
 *
 * @param {ArrayBuffer} arrayBuffer
 * @returns {Array<{ lon: number, lat: number, h: number }>}
 */
export function parseGlb(arrayBuffer) {
  const dv = new DataView(arrayBuffer)

  // Validate GLB magic  'g','l','T','F'
  if (dv.getUint8(0) !== 0x67 || dv.getUint8(1) !== 0x6C ||
      dv.getUint8(2) !== 0x54 || dv.getUint8(3) !== 0x46) {
    console.warn('[glbParser] Not a valid GLB file')
    return []
  }

  // ── Parse JSON chunk ───────────────────────────────────────────────────
  const jsonLen = dv.getUint32(12, true)
  const gltf    = JSON.parse(
    new TextDecoder().decode(new Uint8Array(arrayBuffer, 20, jsonLen))
  )

  // Binary chunk starts after: 12-byte GLB header + 8-byte chunk-0 header + jsonLen
  const binBase = 20 + jsonLen + 8

  // ── Locate POSITION accessor ───────────────────────────────────────────
  let posAccIdx = -1
  for (const mesh of (gltf.meshes || [])) {
    for (const prim of (mesh.primitives || [])) {
      if (prim.attributes?.POSITION !== undefined) {
        posAccIdx = prim.attributes.POSITION
        break
      }
    }
    if (posAccIdx >= 0) break
  }
  if (posAccIdx < 0) {
    console.warn('[glbParser] No POSITION attribute found')
    return []
  }

  const acc    = gltf.accessors[posAccIdx]
  const bv     = gltf.bufferViews[acc.bufferView]
  const count  = acc.count
  // byteStride from bufferView — mago3d uses 8 (3 × uint16 + 2 bytes padding)
  const stride = bv.byteStride || 6
  const binOff = binBase + (bv.byteOffset || 0) + (acc.byteOffset || 0)
  const raw    = new DataView(arrayBuffer, binOff)

  // ── Extract node transforms ────────────────────────────────────────────
  let rtx = 0, rty = 0, rtz = 0   // RootNode translation (glTF Y-up ECEF)
  let pcm = null                    // PointCloudNode matrix (col-major 16 floats)

  for (const nd of (gltf.nodes || [])) {
    if (nd.matrix && nd.mesh !== undefined) {
      // Node with a matrix + mesh reference → PointCloudNode
      pcm = nd.matrix
    } else if (nd.translation && nd.children && !nd.matrix) {
      // Node with translation only + children → RootNode
      ;[rtx, rty, rtz] = nd.translation
    }
  }

  if (!pcm) {
    console.warn('[glbParser] PointCloudNode matrix not found')
    return []
  }

  const M = pcm   // alias for readability

  // ── Parse every point ──────────────────────────────────────────────────
  const out = []
  for (let i = 0; i < count; i++) {
    const base = i * stride

    // Step 1: read uint16, divide by 65535 (normalized accessor)
    const lx = raw.getUint16(base,     true) / 65535
    const ly = raw.getUint16(base + 2, true) / 65535
    const lz = raw.getUint16(base + 4, true) / 65535

    // Step 2: apply PointCloudNode matrix (column-major 4×4)
    const wx = M[0]*lx + M[4]*ly + M[8]*lz  + M[12]
    const wy = M[1]*lx + M[5]*ly + M[9]*lz  + M[13]
    const wz = M[2]*lx + M[6]*ly + M[10]*lz + M[14]

    // Step 3: add RootNode translation
    const gx = wx + rtx
    const gy = wy + rty
    const gz = wz + rtz

    // Step 4: glTF Y-up → ECEF Z-up:  ecef = (gx, -gz, gy)
    out.push(ecefToGeodetic(gx, -gz, gy))
  }

  return out
}

/**
 * Walk a 3D Tiles tile tree and collect all .glb leaf URIs.
 *
 * @param {object} tile   — root tile from tileset.json
 * @param {string[]} out  — accumulator
 */
export function collectGlbUris(tile, out) {
  if (!tile) return
  const uri = tile.content?.uri || tile.content?.url || ''
  if (uri.toLowerCase().endsWith('.glb')) out.push(uri)
  if (tile.children) tile.children.forEach(c => collectGlbUris(c, out))
}

/**
 * Fetch a tileset.json, walk its tree, fetch + parse every .glb tile,
 * and return the combined array of geodetic points.
 *
 * @param {string} tilesetUrl  — absolute or relative URL to tileset.json
 * @returns {Promise<Array<{ lon, lat, h }>>}
 */
export async function loadAllPoints(tilesetUrl) {
  if (!tilesetUrl) return []

  const baseUrl = new URL(tilesetUrl, window.location.href)

  let tsJson
  try {
    const r = await fetch(tilesetUrl)
    if (!r.ok) {
      console.error('[glbParser] tileset fetch failed:', tilesetUrl, r.status)
      return []
    }
    tsJson = await r.json()
  } catch (e) {
    console.error('[glbParser] tileset parse error:', e)
    return []
  }

  const uris = []
  collectGlbUris(tsJson.root, uris)
  console.log(`[glbParser] ${uris.length} GLB tiles in ${tilesetUrl}`)

  const groups = await Promise.all(uris.map(async uri => {
    const url = new URL(uri, baseUrl).href
    try {
      const r = await fetch(url)
      if (!r.ok) {
        console.warn('[glbParser] tile fetch failed:', url, r.status)
        return []
      }
      return parseGlb(await r.arrayBuffer())
    } catch (e) {
      console.warn('[glbParser] tile parse error:', url, e)
      return []
    }
  }))

  const all = groups.flat()
  console.log(`[glbParser] ${all.length} total points from ${tilesetUrl}`)
  return all
}

File: server/glb_parser.py
"""
glb_parser.py — mago3d-tiler GLB point-cloud parser (server-side)

Mirrors the exact transform pipeline from lib/glbParser.js using numpy
for vectorised operations (~10-50x faster than the browser JS version).

Transform pipeline (verified against real Korea site files):
  1.  uint16 / 65535            → normalised local coords  [0, 1]
  2.  Apply PointCloudNode matrix M (column-major 4×4)
  3.  Add RootNode translation
  4.  glTF Y-up → ECEF Z-up:   ecef = (gx, -gz, gy)
  5.  ECEF → (lon°, lat°, h m)  via Bowring iterative method (10 iters)
"""

import json
import struct
from pathlib import Path

import numpy as np

# ── WGS-84 constants ──────────────────────────────────────────────────────
_A  = 6_378_137.0
_E2 = 0.006_694_379_990_14


def _ecef_to_geodetic(ecef: np.ndarray) -> np.ndarray:
    """
    Vectorised ECEF → geodetic (lon°, lat°, h m).

    Parameters
    ----------
    ecef : (N, 3) float64  [x, y, z] metres

    Returns
    -------
    (N, 3) float64  [lon°, lat°, h m]
    """
    x, y, z = ecef[:, 0], ecef[:, 1], ecef[:, 2]
    p   = np.sqrt(x * x + y * y)
    lat = np.arctan2(z, p * (1.0 - _E2))

    for _ in range(10):                          # Bowring iteration
        s   = np.sin(lat)
        N   = _A / np.sqrt(1.0 - _E2 * s * s)
        lat = np.arctan2(z + _E2 * N * s, p)

    s   = np.sin(lat)
    c   = np.cos(lat)
    N   = _A / np.sqrt(1.0 - _E2 * s * s)
    h   = np.where(
        np.abs(c) > 1e-9,
        p / c - N,
        np.abs(z) / s - N * (1.0 - _E2),
    )
    lon = np.arctan2(y, x) * (180.0 / np.pi)

    return np.column_stack([lon, lat * (180.0 / np.pi), h])


def parse_glb(glb_path: Path) -> np.ndarray:
    """
    Parse one mago3d-tiler .glb point-cloud file from disk.

    Parameters
    ----------
    glb_path : Path to the .glb file

    Returns
    -------
    (N, 3) float64  [lon°, lat°, h m]
    Empty (0, 3) array on any error.
    """
    data = glb_path.read_bytes()

    # Validate GLB magic  b'glTF'
    if data[:4] != b"glTF":
        print(f"[glb_parser] Not a valid GLB: {glb_path}")
        return np.zeros((0, 3), dtype=np.float64)

    json_len = struct.unpack_from("<I", data, 12)[0]
    gltf     = json.loads(data[20 : 20 + json_len])

    # Binary chunk starts after: 12-byte GLB header
    #                          + 8-byte chunk-0 header (len + type)
    #                          + json_len bytes
    bin_base = 20 + json_len + 8

    # ── Locate POSITION accessor ──────────────────────────────────────────
    pos_acc_idx = -1
    for mesh in gltf.get("meshes", []):
        for prim in mesh.get("primitives", []):
            if "POSITION" in prim.get("attributes", {}):
                pos_acc_idx = prim["attributes"]["POSITION"]
                break
        if pos_acc_idx >= 0:
            break

    if pos_acc_idx < 0:
        print(f"[glb_parser] No POSITION attribute: {glb_path}")
        return np.zeros((0, 3), dtype=np.float64)

    acc     = gltf["accessors"][pos_acc_idx]
    bv      = gltf["bufferViews"][acc["bufferView"]]
    count   = acc["count"]
    # mago3d uses byteStride = 8  (3 × uint16 = 6 bytes + 2 bytes padding)
    stride  = bv.get("byteStride", 6)
    bin_off = bin_base + bv.get("byteOffset", 0) + acc.get("byteOffset", 0)

    # ── Node transforms ───────────────────────────────────────────────────
    rtx = rty = rtz = 0.0
    pcm = None

    for nd in gltf.get("nodes", []):
        if "matrix" in nd and nd.get("mesh") is not None:
            pcm = nd["matrix"]                           # PointCloudNode
        elif "translation" in nd and "children" in nd and "matrix" not in nd:
            rtx, rty, rtz = nd["translation"]            # RootNode

    if pcm is None:
        print(f"[glb_parser] PointCloudNode matrix not found: {glb_path}")
        return np.zeros((0, 3), dtype=np.float64)

    M = pcm   # 16-element list, column-major

    # ── Read uint16 positions (vectorised) ────────────────────────────────
    # stride in bytes, uint16 = 2 bytes → stride // 2 elements per row
    elems = stride // 2
    raw   = (
        np.frombuffer(
            data[bin_off : bin_off + count * stride], dtype=np.uint16
        )
        .reshape(count, elems)[:, :3]          # keep only x, y, z
        .astype(np.float64) / 65535.0
    )

    lx, ly, lz = raw[:, 0], raw[:, 1], raw[:, 2]

    # ── Apply PointCloudNode matrix (column-major 4×4) ────────────────────
    wx = M[0]*lx + M[4]*ly + M[8]*lz  + M[12]
    wy = M[1]*lx + M[5]*ly + M[9]*lz  + M[13]
    wz = M[2]*lx + M[6]*ly + M[10]*lz + M[14]

    # ── Add RootNode translation ──────────────────────────────────────────
    gx = wx + rtx
    gy = wy + rty
    gz = wz + rtz

    # ── glTF Y-up → ECEF Z-up → geodetic ─────────────────────────────────
    ecef = np.column_stack([gx, -gz, gy])
    return _ecef_to_geodetic(ecef)


def _collect_glb_uris(tile: dict, out: list) -> None:
    """Walk a 3D-Tiles tile tree and collect all .glb content URIs."""
    if not tile:
        return
    content = tile.get("content", {})
    uri = content.get("uri") or content.get("url", "")
    if uri.lower().endswith(".glb"):
        out.append(uri)
    for child in tile.get("children", []):
        _collect_glb_uris(child, out)


def load_all_points(tileset_path: Path) -> np.ndarray:
    """
    Read a tileset.json from disk, walk its tile tree, parse every .glb,
    and return the combined geodetic point array.

    Parameters
    ----------
    tileset_path : absolute Path to tileset.json

    Returns
    -------
    (N, 3) float64  [lon°, lat°, h m]
    """
    if not tileset_path.exists():
        print(f"[glb_parser] tileset not found: {tileset_path}")
        return np.zeros((0, 3), dtype=np.float64)

    with open(tileset_path) as f:
        ts_json = json.load(f)

    uris: list[str] = []
    _collect_glb_uris(ts_json.get("root", {}), uris)
    print(f"[glb_parser] {len(uris)} GLB tiles in {tileset_path}")

    tile_dir = tileset_path.parent
    batches  = []
    for uri in uris:
        glb_path = (tile_dir / uri).resolve()
        pts = parse_glb(glb_path)
        if pts.shape[0] > 0:
            batches.append(pts)

    if not batches:
        return np.zeros((0, 3), dtype=np.float64)

    all_pts = np.vstack(batches)
    print(f"[glb_parser] {all_pts.shape[0]} total points from {tileset_path}")
    return all_pts